### DISCUSSION TITLE
Set foreign key dynamically

### DISCUSSION THREAD

**Discussion by conformist-mw** [sqlalchemy/sqlalchemy#10542] [Usage Questions] [answered]
Hi.
I've already searched the documentation and looked at Generic Associations examples, but I can't find a solution to my problem.
I'm working on a library that embeds into a project and creates its own tables, adding new functionality.
This library has its own models and methods to work with, but to allow it to evolve separately, it is completely decoupled from projects and takes a session object in each method to communicate with the database.

The simplest example of a model and method is this:
```python
# shared_lib/models.py
from sqlalchemy import Column, Integer, String, select, DateTime, func
from sqlalchemy.orm import DeclarativeBase


class SharedModel(DeclarativeBase):
    __abstract__ = True

    created_at = Column(DateTime, server_default=func.now())


class UserFeature(SharedModel):
    __tablename__ = "user_features"
    id = Column(Integer, primary_key=True)
    name = Column(String(20), nullable=False)
    user_id = Column(Integer, nullable=False)


def get_user_features(db_session, user_id):
    return db_session.scalars(
        select(UserFeature).where(UserFeature.user_id == user_id)
    ).all()
```

Accordingly, after installing the library into the project, I import all models inside models.py of the project itself:

```python
# models.py
from sqlalchemy import Column, Integer, String, DateTime, func
from sqlalchemy.orm import DeclarativeBase

from shared_lib.models import UserFeature


class Model(DeclarativeBase):
    __abstract__ = True

    created_at = Column(DateTime, server_default=func.now())


class User(Model):
    __tablename__ = "users"
    id = Column(Integer, primary_key=True)
    name = Column(String(20), nullable=False)
```

I use the metadata of both base models to track changes in all models with alembic:

```python
# alembic/env.py
from shared_lib.models import SharedModel
from models import Model

target_metadata = [SharedModel.metadata, Model.metadata]
```
The problem is that projects are different and the User model can be named differently (User/MainUser/etc.). Therefore, at the moment of UserFeature initialization I cannot specify ForeignKey. 

In general this all works, but without the ForeignKey constraint I can easily do something like this: 
```python
engine = create_engine("sqlite:///example.db")

with Session(engine) as session:
    session.add(UserFeature(name="name", user_id=1000))
    session.commit()
```
user with id=1000 does not exists althouth.

So the question is: can I somehow deferred initialize the user_id or overwrite it on import maybe?

Thanks in advance!

**Comment by CaselIT** [comment]
Hi,

Sure, you can use https://docs.sqlalchemy.org/en/20/core/metadata.html#sqlalchemy.schema.Table.append_constraint to add a fk to an existing table.

You can probably define a library function `register_user_table` to be called in the project with the user table. This function can then create and add the fk constraint to the UserFeature table

**Comment by conformist-mw** [comment]
Thanks for such a quick response!

I tried doing it this way in the test project I took these models from and added them to my question and I get an error:

```python
fk = ForeignKeyConstraint([UserFeature.user_id], ["users.id"])
UserFeature.__table__.append_constraint(fk)
```
```
sqlalchemy.exc.NoReferencedTableError: Foreign key associated with column 'user_features.user_id' could not find table 'users' with which to generate a foreign key to target column 'id'
```

I was also thinking about completely replacing this field, is that possible? For example, if in one of the projects the unique identifier will be not Integer, but UUID, for example.

**Reply by CaselIT** [reply]
From what I understand you have two different metadata, so I think you need to use the actual column, not a string name

**Comment by conformist-mw** [comment]
I tested this on the target database and everything works as expected. Final edits:

```python
class User(Model):
    __tablename__ = "users"
    id = Column(Integer, primary_key=True)
    name = Column(String(20), nullable=False)

    features = relationship(UserFeature)


fk = ForeignKeyConstraint([UserFeature.user_id], [User.id], name="user_feature_user_id_user_id_fc")
UserFeature.__table__.append_constraint(fk)
UserFeature.user = relationship(User, back_populates="features")
```
@CaselIT thank you very much, that really solved the problem.

But still I want to ask, is it possible to change the field type dynamically somehow? In case User.id will be UUID, this option will not work.
Would it help to convert `UserFeature.user_id` to property and depending on the type write to the right column, e.g. `user_integer_id`/`user_uuid_id`? And use `user_id` in query building (e.g. `select(UserFeature).where(UserFeature.user_id == 1)  # <- this is the property`)

**Reply by CaselIT** [reply]
I'm not sure if you can use a property directly. that will probably not work.
You can omit the type in the column if that has a fk when defining a table, 

<details><summary>edit: please disregard the following, see following post</summary>
<p>


but it seems it does not work when adding the fk later:
```py
import sqlalchemy as sa


m = sa.MetaData()
t1 = sa.Table(
    "t",
    m,
    sa.Column("id", sa.Integer, primary_key=True),
    sa.Column("other", sa.Text),
)
t2 = sa.Table(
    "f",
    m,
    sa.Column("id", sa.Integer, primary_key=True),
    sa.Column("fk", sa.ForeignKey(t1.c.id)),
    sa.Column("fk_later"),
)

t2.append_constraint(sa.ForeignKeyConstraint([t1.c.id], [t2.c.fk_later]))

print(repr(t2))
```

I should be safe to do though
```py
t2.c.fk_later.type = t1.c.id.type
```

cc @zzzeek 


</p>
</details>

**Reply by CaselIT** [reply, marked as answer]
You can omit the type in the column if that has a fk when defining a table or when adding it later
```py
import sqlalchemy as sa


m = sa.MetaData()
t1 = sa.Table(
    "t",
    m,
    sa.Column("id", sa.Integer, primary_key=True),
    sa.Column("other", sa.Text),
)
t2 = sa.Table(
    "f",
    m,
    sa.Column("id", sa.Integer, primary_key=True),
    sa.Column("fk", sa.ForeignKey(t1.c.id)),
    sa.Column("fk_later"),
)

print(repr(t2))
t2.append_constraint(sa.ForeignKeyConstraint([t2.c.fk_later], [t1.c.id]))
print(repr(t2))
```

the type of `fk_later` will be set after the constraint it added

**Comment by conformist-mw** [comment]
Interesting. I didn't know that I can omit column type. So I did:

```python
class UserFeature(SharedModel):
    ...
    user_id = Column(nullable=False)

class User(Model):

    id = Column(UUID, primary_key=True)
```
and alembic correctly detect changing the fields types:
```python
def upgrade() -> None:
    op.alter_column('user_features', 'user_id',
               existing_type=sa.INTEGER(),
               type_=sa.UUID(),
               existing_nullable=False)
    op.alter_column('users', 'id',
               existing_type=sa.INTEGER(),
               type_=sa.UUID(),
               existing_nullable=False)
```
Need to test that setup, but looks like it solves all the problems at this moment.

Thank you @CaselIT one more time!
